keps/sig-cloud-provider/2025-extend-konnectivity-for-both-directions/README.md

```diff
@@ -0,0 +1,696 @@
+    - dst_port: The remote port used by the KAS.
+e.g. --target=6443:apiserver.svc.cluster.local:6443
+e.g. --target=6443:[2001:db8:1f70::999:de8:7648:6e8]:6443
+
+* `--bind-address=ip`: Local IP address where the Konnectivity Agent will listen for incoming requests. It will be bound to a dummy interface with IP x.y.z.v defined by the user. Must be used with the previous one to enable incoming requests. If not, and for backward compatibility, only the traffic initiated from the Control Plane will be allowed.
+
+### Handling the Traffic from the pods to the agent
+
+As mentioned above, pods make use of the Kubernetes default service to reach the KAS. To keep things transparent from a pod perspective, they will hit the Konnectivity Agent using the Kubernetes default service. The endpoints will contain the Konnectivity Agent address instead of the KAS one.
+The configuration will be done using the KAS flags `advertise-address ip` and `secure-port port`, that should match respectively the `bind-address` and the `local_port` of the Konnectivity Agent described above.
+
+### Handling the Traffic from the Kubelet to the agent
+
+Kubelet does not use the Kubernetes default service to reach the KAS. Instead it relies on a bootstrap kubeconfig file that is used to connect to the KAS. It then generates a proper kubeconfig file that will be using the same URL.
+Instead of specifying the KAS FQDN/address in the bootstrap kubeconfig file, we will be using the local IP address of the Konnectivity agent (`--bind-address ip`).
+
+### Deployment Model
+
+The agent can be run as static pod or systemd unit. In any case the agent should be started to give access to the KAS and kubelet first and to the hosted pods later. This means that using DaemonSet or Deployment is not an option in this setup, because the kubelet would not be able to get the pod manifests from the KAS.
+Note that no communications between kubelet/pods on the node network and KAS will be possible until the agent is up and running. Therefore the kubelet must handle the absence of connectivity with the KAS properly (the retry logic has to be verified).
+
+### Listening Interface for Konnectivity agent
+
+As mentioned before, we will be using the Kubernetes default service to route traffic to the agent. The service in itself has a couple of limitations: it can’t be used as a type externalName, thus preventing usage of DNS names. But also, some general services limitations apply: endpoints can't use the link-local range and the localhost range. This means that we are left with the 3 private IPs ranges (10.0.0.0/8, 172.16.0.0/12 and 192.168.0.0/16).
+
+The agent will create a dummy interface, assign it the ip provided with the `bind-address` flag, using `host` scope, and will start listening on this IP:local_port (local_port is defined with the `target` flag). This will allow all agents to bind to the ip address advertised by the KAS, that will be valid only inside the node.
```

Comment:
Not sure if I understand the load-balancer scenario correctly.
If you set the KAS LB IP via `advertise-address` to set it as endpoint for the `kubernetes` svc, you wouldn't need the agents in the first place, would you?

On the LB side there's another potential downside for providers that try to host multiple KAS instances from different clusters behind one LB. There's no way to differentiate to which cluster the traffic belongs. SNI information will yield `kubernetes` for all. The SNI info from the agent connections makes that possible.

Just using the load-balancer approach on the master could also lead to a problem. IIRC on some cloud providers services behind the LB cannot directly talk to the external LB IP. You'd need to "dnat" this again on the master node, which we could do anyway (also in other scenarios) on the master node.

Personally I'd be fine to not be able to use the kubernetes svc on the master and be forced to configure the services with the k8s endpoint.

With regards to network ranges that we could use: Shouldn't we also be able to make use of the "carrier grade NAT" range (100.64.0.0 - 100.127.255.255). I'm assuming this has less chance of conflicting with any existing cluster ranges.

Reply:
Hi @gottwald, thanks for reacting!

> If you set the KAS LB IP via `advertise-address` to set it as endpoint for the `kubernetes` svc, you wouldn't need the agents in the first place, would you?

I think I did not express clearly enough my idea. What I meant is that we could use a private LB with an IP that is not routable from the node network, and will be used from within the master network only, to allow the PODs to reach the KAS via the default kubernetes service. On the node network, things would be unchanged.

> Personally I'd be fine to not be able to use the kubernetes svc on the master and be forced to configure the services with the k8s endpoint.

If no one has something against it, I would be fine to start with this limitation too.

> With regards to network ranges that we could use: Shouldn't we also be able to make use of the "carrier grade NAT" range (100.64.0.0 - 100.127.255.255). I'm assuming this has less chance of conflicting with any existing cluster ranges.

Thx for the hint. I did not know about this range. Using link-local range would be probably the safest/cleanest solution, but as we cannot because of endpoints limitations I have nothing against mentioning this range too, @youssefazrak Any thoughts about this?

Reply:
Nothing against it. And actually, that's a good idea. 
The range is private to the cluster and supposedly a host reaches from a CGN network, there will be no routing conflict as we are NAT to another range.